draft-nygren-httpbis-httpssvc.md

```diff
@@ -0,0 +1,791 @@
+  accessing the origin in the absence of Alt-Svc.
+* An "alternative service" is a different server that can serve the
+  origin.
+
+Abstractly, the origin consists of a scheme (typically "https"), a host
+name, and a port (typically "443").
+
+
+Additional DNS terminology intends to be consistent
+with {{?DNSTerm=RFC8499}}.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
+NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+"MAY", and "OPTIONAL" in this document are to be interpreted as
+described in BCP 14 {{!RFC2119}} {{!RFC8174}} when, and only when, they
+appear in all capitals, as shown here.
+
+
+
+# The HTTPSSVC record type
+
+The HTTPSSVC DNS resource record (RR) type (RRTYPE ???)
+is used to locate endpoints that can service an "https" origin.
+The presentation format of the record is:
+
+ RRName TTL Class HTTPSSVC SvcRecordType SvcDomainName SvcFieldValue
+
+where SvcRecordType is a numeric value of either 0 or 1,
+SvcDomainName is a domain name, and SvcFieldValue is a string
+present when SvcRecordType is 1.
+
+The algorithm for resolving HTTPSSVC records and associated
+address records is specified in TBD:WriteMe.
+
+
+## HTTPSSVC RDATA Wire Format
+
+The RDATA for the HTTPSSVC RR consists of:
+
+* a 1 octet flag field for SvcRecordType, interpreted
+  as an unsigned numeric value (0 to 255, with only values
+  "0" and "1" defined here)
+* a 1 octet length field for the SvcDomainName
+* a 2 octet length field for the SvcFieldValue
+* the uncompressed SvcDomainName byte string
+  of the specified length (up to 253 characters)
+* the SvcFieldValue byte string of the specified
+  length (up to 65536 characters)
+
+When SvcRecordType is 0, the length of SvcFieldValue SHOULD be 0
+and clients MUST ignore the contents of non-empty SvcFieldValue fields.
+
+
+## RRNames
+
+In the case of the HTTPSSVC RR, an origin is translated into the RRName
+in the following manner:
+
+1. The origin scheme MUST be "https".  If the port is unspecified or "443",
+   then the RRName is equal to the origin host name.
+
+2. If the port is not 443, the RRName is represented by prefixing the
+   port and scheme with "_", then concatenating them with the host,
+   resulting in a domain name like "_832._https.www.example.com.".
+
+3. When a prior CNAME or HTTPSSVC record has aliased to
+   an HTTPSSVC record, RRName shall be the name of the alias target.
+
+Note that none of these forms alter the HTTPS origin or authority.
+For example, clients MUST continue to validate TLS certificate
+hostnames based on the origin host.
+
+
+## SvcRecordType
+
+The SvcRecordType field is a numeric value defined to be either "0" or
+"1". Within an HTTPSSVC RRSet, all RRs must have the same value for
+SvcRecordType.  Clients and recursive servers MUST ignore HTTPSSVC
+resource records with other SvcRecordType values, as well as
+RRSets where both "0" and "1" values for SvcRecordType are present in
+the same RRSet.
```

Comment:
Good idea...  At least keeps it deterministic.

Reply:
What is the advantage of preferring version 0?  Intuitively, I would expect a preference for version 1, since it enables better performance and security.  For future-proofing, I could imagine a rule like "prefer the highest type number you support, and ignore anything higher".

Reply:
Two reasons, but I could be persuaded otherwise:  There's a requirement that there be only one "0."  If the rule is "consume the first zero, ignore anything else" then you have an easy and deterministic rule.  Also, it makes it easy to roll out "1" by populating the new records prior to removing the old one.

And of course, something is better than nothing -- I'm amenable to other deterministic ways of selecting, but I'm less attracted to "pretend you didn't see it" as a mandated reaction.

Reply:
I think what Mike says makes sense (to use one of the "0"'s, ignore the others if multiple 0's or a mix of 0/1 are present).  These aren't different versions/types, they have different semantics.  (The "0" is a CNAME/ANAME equivalent and the "1" is the ALTSVC record.)  I guess we could swap the 0/1 value definitions and then say "use the highest you understand" or we could also just say to ignore ones you don't understand and then let any future draft that defines "2" if-any tell clients how to prefer or disprefer it.)

Reply:
OK, proposed text in #16